File: fibrb/fib.rb
require 'thread'
require 'rubygems'
require 'agent'

def produce chan, limit 
  a, b = 0, 1
  chan << 0
  while b < limit
    chan << b
    print "Producer sent #{b}\n"
    a, b = b, a+b
  end
  chan.close
end

def consume chan, done, n
  until chan.closed?
    sleep(rand(100) * 0.001)
    i = chan.receive
    print "Consumer #{n} received #{i}\n"
  end
rescue
ensure
  done << 1
end

chan = Agent::Channel.new(name: 'fib', type: Integer)
done = Agent::Channel.new(name: 'complete', type: Integer)
go{produce chan, 1000}
10.times{|i| go{consume(chan, done, i)}}
10.times{done.receive}

File: fib/fib.go
package main

import (
	"fmt"
	"time"
	"rand"
)

var done = make(chan int, 10)

func main() {
	c := make(chan int)
	go produce(c, 1000)
	// start consumers
	for i := 0; i < 10; i++ {
		go consume(c, i)
	}
	// wait for consumers to finish
	for i := 0; i < 10; i++ {
		<-done
	}
}

func produce(c chan int, until int) {
	var a,b int
	c <- 0
	// c <- 1
	for a, b = 0, 1; b < until; a, b = b, a+b {
		c <- b
	}
	close(c)
}

func consume(c chan int, n int) {
	for {
		time.Sleep(rand.Int63n(100) * 1e6)
		if i, ok := <-c; ok {
			fmt.Printf("Consumer %v consumed %v\n", n, i)
		} else {
			done <-1
			return
		}
	}
}